### PR TITLE
Adding the basic support for diktat and detekt linters

### PR DESCRIPTION
**Pull request by akuleshov7**
Follow up of the following issue: #65

### What's done:
- Added gradle tasks to run the latest [diktat](https://github.com/diktat-static-analysis/diKTat/issues) and [detekt](https://github.com/detekt/detekt) static analyzers to the project. These linters together contain ktlint standard formatting rules + an overhead with more than 200 inspections for the code;
- Added corresponding CI/CD tasks to support continuous code style checking;
- Refactored the code to make diktat and build/test tasks passing 

### What's left:
- As the code is not consistent, to avoid huge PR warnings raised by detekt are not fixed yet, should be done in phaze#2
- Following Diktat warnings are globally suppressed for the project:  MISSING_KDOC_CLASS_ELEMENTS, MISSING_KDOC_TOP_LEVEL, HEADER_MISSING_IN_NON_SINGLE_CLASS_FILE, VARIABLE_HAS_PREFIX, AVOID_USING_UTILITY_CLASS, CUSTOM_GETTERS_SETTERS. These rules (except related to Kdocs) are controvertial and require a separate PR.

### CODE REVIEW

File: reflekt-plugin/src/main/kotlin/io/reflekt/plugin/analysis/ir/ArgumentCollectors.kt
Comment (nbirillo) [inline reflekt-plugin/src/main/kotlin/io/reflekt/plugin/analysis/ir/ArgumentCollectors.kt:39]: Are you sure about using these empty blocks? I mean it could complicate the source code. Maybe it is better to disable this inspection?

File: CONTRIBUTING.md
Comment (nbirillo) [inline CONTRIBUTING.md:1]: I am not sure that is helpful for contributors. I mean you should add a subheader about code style checking for example and a todo about contribution information. I will fill it in later

**Review by nbirillo — CHANGES_REQUESTED** [review]
Great work! Thank you a lot for your PR since it is very helpful for us! I have several small comments, also I have an additional question.

You did not fix detekt issues (you commented on it in the description). Are you going to fix these issues in the next PR or should we fix all of them by ourselves? 

Additionally, a question that is not related to solving this issue. The main code is understandable for you? Did you have some troubles with a local run and build? I mean, we will try to make the contribution process easier and your feedback could be very useful for us.

**Comment by akuleshov7** [comment]
Hi, Nastya! @nbirillo 

thank you for your response. Let me answer to you general questions:

> You did not fix detekt issues (you commented on it in the description). Are you going to fix these issues in the next PR or should we fix all of them by ourselves?

Yes, I do not think that you will be physically able to review a PR, that has more changes 😄. So I [decided](https://github.com/JetBrains-Research/reflekt/issues/65#issuecomment-965725790) to split them in two PRs.  As some of diktat inspections have automated fixer, and diktat is more [focused on the coding style](https://github.com/diktat-static-analysis/diKTat/blob/master/info/guide/diktat-coding-convention.md), I decided to start from the easiest part. I have not touched any of detekt issues right now, but will fix them in phaze#2.

**Comment by akuleshov7** [comment]
> Additionally, a question that is not related to solving this issue. The main code is understandable for you? Did you have some troubles with a local run and build? I mean, we will try to make the contribution process easier and your feedback could be very useful for us.

Would you like to get a bad answer or a good one? 🤣  Once I got a [feedback to my project](https://github.com/akuleshov7/ktoml/issues/49) that completely destroyed me. Would you like to have the the same? 😆
 
I will try to be fair: I have read all the project, while I was rewriting mostly all the code and the code itself was not so good as the scientific part that you are doing ❤️

There are a lot of things that show that you guys are more Java, than a Kotlin developers. For example - you have a bunch of utility methods wrapped in objects, that can be converted into beautiful top-level extension functions. 

There is a lot of code smells, e.g.: code duplication, unused methods, documentation is missing also in places where it is needed. That will be fixed (I guess) after we will integrate detekt and diktat and create a proper CI/CD for it.

But the DSL and the basic usage of the code is good enough, I love this thing.
There were also no problems to running the build and tests, but I would actually rework your test framework and resources, right now it's quite difficult to debug it.

File: CONTRIBUTING.md
Comment (akuleshov7) [reply CONTRIBUTING.md:1]: I am afrad that a lot of people do not know how to run static analysis tasks in gradle. And my implementation with buildsrc looks beautiful, but not very known for amature developers. 

I beat that after a lot of developers will try to contribute to you, they would have a problem with that :)

File: reflekt-plugin/src/main/kotlin/io/reflekt/plugin/analysis/ir/ArgumentCollectors.kt
Comment (akuleshov7) [reply reflekt-plugin/src/main/kotlin/io/reflekt/plugin/analysis/ir/ArgumentCollectors.kt:39]: Sure, that can be suppressed. That was automatically generated to attract your attention.
I have seen several bugs realted to a missing else block in the code, but I also do not like [this inspection](https://github.com/diktat-static-analysis/diKTat/blob/master/info/guide/diktat-coding-convention.md#-311-when-expression)

File: CONTRIBUTING.md
Comment (nbirillo) [reply CONTRIBUTING.md:1]: Wow, I did not know it. I thought it is obvious and everyone can run analyzers with Gradle. Then thank you!

Initially, we did not add analyzers because we had a lot of different attempts and approaches to make Reflekt and it was lost between them.

**Comment by nbirillo** [comment]
@akuleshov7 Thank you a lot for your feedback! 

For me, it is good to get a bad answer, since then we can improve the plugin and make it better. 

About objects for utils - yeah, it is a problem that we did not solve everywhere since we wanted firstly to add static analyzers and then refactor the code. I don't know if you saw the open PRs, but we made a lot of changes to improve the main logic and fix the work with libraries. The next stage is supported incremental compilation (we can not do it immediately since the API does not have the necessary functions, but we are in progress), code refactoring and extending its DSL. I had a small vacation and did not have time to finish all these issues before.

About code smells - yeah, I also know about this is also connected to constantly rewriting Reflekt code and should be deleted in the future. Of course, it can be found easily by using Detekt (or something like this)

I am very happy that you found DSL and the basic usage of the code good as well as the research part! Thank you a lot! ❤️

It is great that you are going to fix Detekt issues. But I am afraid we have big trouble. 🥲 @elena-lyulina should check my open PRs at the beginning of the next week and all of them contain A LOT of changes since I added some tests architecture for incremental compilation, rewritten library handlers part and did some small refactorings. But the main big change is package renaming that we need for publishing into the Gradle portal to make the installation process easier. I mean, if you make another big PR we will not be able to merge all changes since it will be very difficult and have a lot of conflicts. But maybe can I merge my changes into your branch and you will fix Detekt issues there? Or can you wait a little bit while we don't merge my changes to merge yours and work with the new Reflekt version? 
We are really sorry about this situation, but we did not expect any participants to appear at this stage (but thank you very much for that)

**Comment by akuleshov7** [comment]
> But maybe can I merge my changes into your branch and you will fix Detekt issues there? Or can you wait a little bit while we don't merge my changes to merge yours and work with the new Reflekt version? We are really sorry about this situation, but we did not expect any contributors to appear at this stage (but thank you very much for that)

Yes, sure, that is not a problem. That is not a problem, I can do changes related to detect AFTER all your pull requests including this one will be merged.

The only thing that I am afraid about is the following: I am afraid that after your merges, when I will be trying to resolve conflicts, I will break the logic that I don’t understand… So may be we need to merge this PR first…

**Comment by akuleshov7** [comment]
But actually, you can do everything in the easiest way for you and I will adjust to you. No problem 😄

**Comment by nbirillo** [comment]
@akuleshov7 Thank you! I changed the target branch (it is a copy of the main branch). In this case, we can merge your PR first 

Could you please fix the comments? Then I will approve it and will merge

File: reflekt-dsl/src/main/kotlin/io/reflekt/util/TypeStringRepresentationUtil.kt
Comment (akuleshov7) [reply reflekt-dsl/src/main/kotlin/io/reflekt/util/TypeStringRepresentationUtil.kt:53]: Actually I do not see usages of this function and it is internal, are you sure that you use it in the Reflekt DSL?
I would rather delete it...

File: gradle-plugin/src/main/kotlin/io/reflekt/plugin/ReflektSubPlugin.kt
Comment (akuleshov7) [inline gradle-plugin/src/main/kotlin/io/reflekt/plugin/ReflektSubPlugin.kt:67]: Please also keep an eye on this inspection. As Kotlin has special language specific null-safe logic, all  Java-like null comparissons are converted into corresponding (`let`/`run`/etc.). If you don't like this one, I will supress it also,but actually I suppose that these features should be used as you are writting Kotlin code, but not Java code.

**Comment by akuleshov7** [comment]
review notes are fixed, please unlock the workflow. You can also make these workflow `required` after we would fix detekt issues

File: reflekt-dsl/src/main/kotlin/io/reflekt/util/TypeStringRepresentationUtil.kt
Comment (nbirillo) [reply reflekt-dsl/src/main/kotlin/io/reflekt/util/TypeStringRepresentationUtil.kt:53]: Sorry, I really meant the one above (`fun KType.stringRepresentation()`) (but it has the same description)

Comment (akuleshov7) [reply reflekt-dsl/src/main/kotlin/io/reflekt/util/TypeStringRepresentationUtil.kt:53]: Let's add the documentation after we will enable diktat inspections that force to write proper Kdocs?

Comment (nbirillo) [reply reflekt-dsl/src/main/kotlin/io/reflekt/util/TypeStringRepresentationUtil.kt:53]: Great idea, thank you!

**Review by nbirillo — APPROVED** [review]
Great! Thank you a lot! We will mention you under the issue after merging all our PRs?

**Comment by nbirillo** [comment]
Phase#1 from the [issue](https://github.com/JetBrains-Research/reflekt/issues/65) is done. Thank you a lot @akuleshov7  🥳

**Pull request merged**
